libgalois/include/katana/analytics/ClusteringImplementationBase.h

```diff
@@ -0,0 +1,680 @@
+
+namespace katana::analytics {
+
+// Maintain community information
+template <typename EdgeWeightType>
+struct CommunityType {
+  std::atomic<uint64_t> size;
+  std::atomic<EdgeWeightType> degree_wt;
+  EdgeWeightType internal_edge_wt;
+};
+
+typedef katana::LargeArray<uint64_t> LargeArray;
+
+using LCPreviousCommunityId = katana::PODProperty<uint64_t>;
+using LCCurrentCommunityId = katana::PODProperty<uint64_t>;
+using LCColorId = katana::PODProperty<int64_t>;
+template <typename EdgeWeightType>
+using LCDegreeWeight = katana::PODProperty<EdgeWeightType>;
+
+template <typename EdgeWeightType>
+using LCEdgeWeight = katana::PODProperty<EdgeWeightType>;
+
+template <typename _Graph, typename _EdgeType, typename _CommunityType>
+struct ClusteringImplementationBase {
+  using Graph = _Graph;
+  using GNode = typename Graph::Node;
+  using EdgeTy = _EdgeType;
+  using CommunityType = _CommunityType;
+
+  /*
+ * Typedefs
+ */
+  constexpr static const uint64_t INF_VAL =
+      std::numeric_limits<uint64_t>::max() / 2 - 1;
+  constexpr static const uint64_t UNASSIGNED =
+      std::numeric_limits<uint64_t>::max();
+  constexpr static const double DOUBLE_MAX =
+      std::numeric_limits<double>::max() / 4;
+
+  typedef katana::LargeArray<uint64_t> LargeArray;
+  typedef katana::LargeArray<EdgeTy> largeArrayEdgeTy;
+
+  typedef katana::LargeArray<CommunityType> CommArray;
+
+  static void PrintGraphCharateristics(const Graph& graph) {
+    katana::gPrint("/******************************************/\n");
+    katana::gPrint("/************ Graph Properties ************/\n");
+    katana::gPrint("/******************************************/\n");
+    katana::gPrint("Number of Nodes: ", graph.num_nodes(), "\n");
+    katana::gPrint("Number of Edges: ", graph.num_edges(), "\n");
+  }
+
+  static void PrintGraphCharateristics(const katana::PropertyGraph& pfg) {
+    katana::gPrint("/******************************************/\n");
+    katana::gPrint("/************ Graph Properties ************/\n");
+    katana::gPrint("/******************************************/\n");
+    katana::gPrint("Number of Nodes: ", pfg.num_nodes(), "\n");
+    katana::gPrint("Number of Edges: ", pfg.num_edges(), "\n");
+  }
+
+  /**
+ * Algorithm to find the best cluster for the node
+ * to move to among its neighbors.
+ */
+  template <typename EdgeWeightType>
+  void FindNeighboringClusters(
+      const Graph& graph, GNode& n,
+      std::map<uint64_t, uint64_t>& cluster_local_map,
+      std::vector<EdgeTy>& counter, EdgeTy& self_loop_wt) {
+    uint64_t num_unique_clusters = 0;
+    /**
+   * Add the node's current cluster to be considered
+   * for movement as well
+   */
+    cluster_local_map[graph.template GetData<LCCurrentCommunityId>(n)] =
+        0;                 // Add n's current cluster
+    counter.push_back(0);  // Initialize the counter to zero (no edges incident
+                           // yet)
+    num_unique_clusters++;
+
```

Comment:
we can skip this. The for loop can handle this. For example, if there is an edge from n to a node m (or itself) in its own community, then the node's current cluster/community will get added to the cluster map.

Reply:
I don't understand what you mean.

Reply:
I was just saying that since the entry is zero, we don't need to add n's communtiy to the map at the start. We can do that inside the for loop (when we find an edge from n to its community)

Reply:
we can leave it like this as well, it's fine.